File: services/elitea-main/internal/api/v2/toolkits/tool_groups.go
package toolkits

import "strings"

// Tool GROUPS: the served, per-type answer to "what does this tool do to the
// target system?"
//
// ─────────────────────────────────────────────────────────────────────────────
// WHY THE SERVER ANSWERS THIS AT ALL
// ─────────────────────────────────────────────────────────────────────────────
// The toolkit editor renders one flat chip list of every tool a type offers.
// For GitHub that is 26+ chips in one row, with `delete_file` sitting between
// `create_file` and `get_file_metadata` and nothing distinguishing them. The
// person ticking those boxes is deciding what an autonomous agent will be
// allowed to do, and the screen gives them no way to see that one of the boxes
// is destructive.
//
// The grouping could have been invented in the web client. It is here instead
// because the tool list is served from here — `properties.selected_tools`
// (`args_schemas` keys, or the older `items.enum`) is this handler's contract
// with the client — and a classification that lived only in the browser would
// have to be re-derived by every other consumer of the same catalogue, and
// would drift the moment the SDK adds a tool.
//
// THE STORED SHAPE IS UNCHANGED. `selected_tools` remains the same array of
// tool NAMES it has always been. Nothing here is written to a toolkit row; an
// agent, a pipeline or an MCP client reading a saved toolkit sees exactly what
// it saw before.
//
// ─────────────────────────────────────────────────────────────────────────────
// WHY THE CLASSIFICATION IS DERIVED FROM THE NAME
// ─────────────────────────────────────────────────────────────────────────────
// The pinned SDK snapshot carries no category, capability or side-effect
// annotation of any kind — its per-tool content is the argument schema and
// nothing else (verified across all 53 entries of
// runtimecomposition/current_toolkit_schema_snapshot.json). So there is no
// upstream field to serve, and the choice is between serving nothing and
// deriving something from the one signal that exists: the tool's name.
//
// The SDK names tools verb-first and consistently — 161 `get_*`, 75 `create_*`,
// 66 `list_*`, 46 `update_*`, 36 `search_*`, 26 `delete_*` across 444 distinct
// tools — so the derivation is reliable in practice, and it is deliberately
// CONSERVATIVE where it is not: the FIRST recognised verb token wins, which is
// what keeps `search_index` a read (it searches an index) while `index_data` is
// a write (it indexes data). A last-token or "any token" rule gets that pair
// backwards.
//
// A tool whose name carries no recognised verb (`places`, `fields_metadata`,
// `job_stats`) falls to Read. That is the correct answer for every such name in
// the pinned snapshot, and the fallback is stated here rather than left
// implicit because it is the one place this derivation could be wrong about a
// future tool: a new SDK tool with an unrecognised WRITE verb would be grouped
// as a read. The group is a display aid over a list the user still ticks
// individually — it never grants or withholds anything — so the failure mode is
// a mislabelled chip, not an unintended permission.

// ToolGroup is the served group identifier. The web client maps these to the
// group's label, badge and tooltip; keeping the identifiers opaque and stable
// is what lets the wording change without a server release.
type ToolGroup string

const (
	// ToolGroupRead returns data. Nothing is created, changed or destroyed.
	ToolGroupRead ToolGroup = "read"
	// ToolGroupCreateUpdate creates or modifies data in the target system.
	ToolGroupCreateUpdate ToolGroup = "create_update"
	// ToolGroupDelete destroys data. Not reversible from this platform.
	ToolGroupDelete ToolGroup = "delete"
	// ToolGroupExecute runs a caller-supplied query, script, pipeline or raw
	// API call. The effect is not bounded by the tool.
	ToolGroupExecute ToolGroup = "execute"
)

// toolGroupOrder is the FIXED display order, least to most dangerous. It is
// served rather than left to the client so every surface that renders these
// groups orders them identically.
var toolGroupOrder = []any{
	string(ToolGroupRead),
	string(ToolGroupCreateUpdate),
	string(ToolGroupDelete),
	string(ToolGroupExecute),
}

// toolGroupVerbs maps a name token to the group it decides.
//
// Read verbs ARE listed even though Read is also the fallback, and that is
// load-bearing rather than redundant: the first RECOGNISED token wins, so
// `search` has to be recognised for `search_index` to come out a read instead
// of being decided by the `index` token behind it.
var toolGroupVerbs = map[string]ToolGroup{
	// Returns data. Nothing is created, changed or destroyed.
	"get": ToolGroupRead, "list": ToolGroupRead, "read": ToolGroupRead,
	"search": ToolGroupRead, "find": ToolGroupRead, "grep": ToolGroupRead,
	"stepback": ToolGroupRead, "similarity": ToolGroupRead, "describe": ToolGroupRead,
	"show": ToolGroupRead, "download": ToolGroupRead, "browse": ToolGroupRead,
	"retrieve": ToolGroupRead, "fetch": ToolGroupRead, "view": ToolGroupRead,
	"extract": ToolGroupRead, "analyze": ToolGroupRead, "metadata": ToolGroupRead,
	"stats": ToolGroupRead, "healthcheck": ToolGroupRead, "count": ToolGroupRead,
	"preview": ToolGroupRead, "export": ToolGroupRead, "diff": ToolGroupRead,

	// Destroys data.
	"delete": ToolGroupDelete, "remove": ToolGroupDelete, "drop": ToolGroupDelete,
	"purge": ToolGroupDelete, "destroy": ToolGroupDelete, "erase": ToolGroupDelete,
	"truncate": ToolGroupDelete, "clear": ToolGroupDelete, "unlink": ToolGroupDelete,
	"uninstall": ToolGroupDelete, "revoke": ToolGroupDelete,

	// Runs something the caller supplied. Not bounded by the tool.
	"execute": ToolGroupExecute, "exec": ToolGroupExecute, "run": ToolGroupExecute,
	"invoke": ToolGroupExecute, "eval": ToolGroupExecute, "generic": ToolGroupExecute,
	"sandbox": ToolGroupExecute, "trigger": ToolGroupExecute, "raw": ToolGroupExecute,

	// Creates or modifies.
	"create": ToolGroupCreateUpdate, "add": ToolGroupCreateUpdate, "update": ToolGroupCreateUpdate,
	"set": ToolGroupCreateUpdate, "write": ToolGroupCreateUpdate, "upload": ToolGroupCreateUpdate,
	"append": ToolGroupCreateUpdate, "patch": ToolGroupCreateUpdate, "edit": ToolGroupCreateUpdate,
	"modify": ToolGroupCreateUpdate, "post": ToolGroupCreateUpdate, "put": ToolGroupCreateUpdate,
	"insert": ToolGroupCreateUpdate, "upsert": ToolGroupCreateUpdate, "send": ToolGroupCreateUpdate,
	"attach": ToolGroupCreateUpdate, "link": ToolGroupCreateUpdate, "move": ToolGroupCreateUpdate,
	"copy": ToolGroupCreateUpdate, "rename": ToolGroupCreateUpdate, "assign": ToolGroupCreateUpdate,
	"apply": ToolGroupCreateUpdate, "index": ToolGroupCreateUpdate, "import": ToolGroupCreateUpdate,
	"publish": ToolGroupCreateUpdate, "comment": ToolGroupCreateUpdate, "submit": ToolGroupCreateUpdate,
	"merge": ToolGroupCreateUpdate, "push": ToolGroupCreateUpdate, "save": ToolGroupCreateUpdate,
	"replace": ToolGroupCreateUpdate, "mark": ToolGroupCreateUpdate, "enable": ToolGroupCreateUpdate,
	"disable": ToolGroupCreateUpdate, "fork": ToolGroupCreateUpdate, "transition": ToolGroupCreateUpdate,
	"register": ToolGroupCreateUpdate, "restore": ToolGroupCreateUpdate, "duplicate": ToolGroupCreateUpdate,
	"checkout": ToolGroupCreateUpdate, "manage": ToolGroupCreateUpdate, "confirm": ToolGroupCreateUpdate,
	"invite": ToolGroupCreateUpdate, "fill": ToolGroupCreateUpdate, "commit": ToolGroupCreateUpdate,
	"close": ToolGroupCreateUpdate, "cancel": ToolGroupCreateUpdate, "generate": ToolGroupCreateUpdate,
	"sync": ToolGroupCreateUpdate,
}

// ClassifyToolGroup answers which group one tool name belongs to.
//
// Exported for the tests that pin the whole pinned-snapshot catalogue against
// it; nothing outside this package calls it in production.
func ClassifyToolGroup(name string) ToolGroup {
	for _, token := range toolNameTokens(name) {
		if group, found := toolGroupVerbs[token]; found {
			return group
		}
	}
	return ToolGroupRead
}

// toolNameTokens splits a tool name into lower-case words, in order.
//
// Both spellings the SDK uses are handled: `snake_case` (the overwhelming
// majority) and the handful of `camelCase` names (`indexDocuments`,
// `stepbackSearch`). A camel boundary that splits `indexDocuments` into
// `index`+`documents` is what keeps that one classified with its snake_case
// twin `index_data` instead of falling to the read fallback.
func toolNameTokens(name string) []string {
	tokens := make([]string, 0, 4)
	current := strings.Builder{}
	flush := func() {
		if current.Len() > 0 {
			tokens = append(tokens, current.String())
			current.Reset()
		}
	}
	for _, character := range name {
		switch {
		case character == '_' || character == '-' || character == '.' || character == ' ':
			flush()
		case character >= 'A' && character <= 'Z':
			flush()
			current.WriteRune(character - 'A' + 'a')
		default:
			current.WriteRune(character)
		}
	}
	flush()
	return tokens
}

// withToolGroups copies one type's settings schema with
// `properties.selected_tools.tool_groups` and `.tool_group_order` added.
//
// It is a no-op for a type whose tool list is not known at catalogue time —
// a Remote MCP toolkit's tools are discovered from the live server and
// arrive in the instance's own `settings.available_mcp_tools`, never here. A
// type with no groups is exactly how the client decides to keep rendering the
// flat list it always did, so "no answer" and "everything is a read" must not
// look the same.
func withToolGroups(settingsSchema map[string]any) map[string]any {
	settingsProperties, _ := settingsSchema["properties"].(map[string]any)
	selectedTools, _ := settingsProperties["selected_tools"].(map[string]any)
	names := toolNamesFromSelectedTools(selectedTools)
	if len(names) == 0 {
		return settingsSchema
	}

	groups := make(map[string]any, len(names))
	for _, name := range names {
		groups[name] = string(ClassifyToolGroup(name))
	}

	// Every node on the written path is rebuilt: toolkitTypeSchemas is
	// package-level state shared by every request, and an in-place edit would
	// leak into the next one — the same rule toolkitTypeCatalogue states.
	rebuiltSelectedTools := make(map[string]any, len(selectedTools)+2)
	for key, value := range selectedTools {
		rebuiltSelectedTools[key] = value
	}
	rebuiltSelectedTools["tool_groups"] = groups
	rebuiltSelectedTools["tool_group_order"] = append([]any(nil), toolGroupOrder...)

	rebuiltProperties := make(map[string]any, len(settingsProperties))
	for key, value := range settingsProperties {
		rebuiltProperties[key] = value
	}
	rebuiltProperties["selected_tools"] = rebuiltSelectedTools

	rebuilt := make(map[string]any, len(settingsSchema))
	for key, value := range settingsSchema {
		rebuilt[key] = value
	}
	rebuilt["properties"] = rebuiltProperties
	return rebuilt
}

// toolNamesFromSelectedTools reads the tool names off a served
// `selected_tools` node: the `args_schemas` keys, or the older `items.enum`.
//
// Exactly the lookup the web client performs
// (`ToolBase.render.tsx`'s resolveAvailableTools and
// `lib/helpers/indexesTabVisibility.ts`), so the group map can never be keyed
// on names the chip list does not render.
func toolNamesFromSelectedTools(selectedTools map[string]any) []string {
	if argsSchemas, ok := selectedTools["args_schemas"].(map[string]map[string]any); ok && len(argsSchemas) > 0 {
		names := make([]string, 0, len(argsSchemas))
		for name := range argsSchemas {
			names = append(names, name)
		}
		return names
	}
	if argsSchemas, ok := selectedTools["args_schemas"].(map[string]any); ok && len(argsSchemas) > 0 {
		names := make([]string, 0, len(argsSchemas))
		for name := range argsSchemas {
			names = append(names, name)
		}
		return names
	}
	items, _ := selectedTools["items"].(map[string]any)
	enum, _ := items["enum"].([]any)
	names := make([]string, 0, len(enum))
	for _, value := range enum {
		if name, ok := value.(string); ok && name != "" {
			names = append(names, name)
		}
	}
	return names
}

File: apps/elitea-web/src/features/toolkits/ui/form/ToolBase/toolGroups.ts
import { t } from '@/shared/i18n';

import type { ToolActionOption } from './ToolActionsItems';

/**
 * Tool GROUPS in the toolkit editor's Tools section — the display side of the
 * classification elitea-main serves at
 * `properties.selected_tools.tool_groups`
 * (`internal/api/v2/toolkits/tool_groups.go`).
 *
 * WHY THE SECTION IS GROUPED AT ALL. The picker used to be one flat chip row.
 * For GitHub that is 26+ chips in a single wrap, with `delete_file` sitting
 * between `create_file` and `get_file_metadata` and nothing telling them
 * apart. The person ticking those chips is deciding what an autonomous agent
 * will be permitted to do, and the screen gave them no way to see that one of
 * the chips was destructive.
 *
 * THE STORED VALUE IS UNCHANGED. `settings.selected_tools` is still the same
 * flat array of tool names it has always been; grouping is a view over it. An
 * agent, a pipeline, or an MCP client reading a saved toolkit sees exactly
 * what it saw before — which is why `ToolActionsSelector` can fall back to the
 * flat list for a toolkit whose tools carry no group metadata, with no
 * migration of anything.
 *
 * THE LABELS LIVE HERE, NOT ON THE SERVER. elitea-main serves opaque group
 * IDs and the fixed order; the wording is this layer's, so it can be
 * translated and reworded without a server release.
 */

/** The fixed order, least to most dangerous. Only the fallback — the served `tool_group_order` wins when present. */
const TOOL_GROUP_ORDER = ['read', 'create_update', 'delete', 'execute'] as const;

export interface ToolGroupDescriptor {
  /** The group's heading. */
  readonly label: string;
  /** The one-word consequence, shown as a badge beside the heading. */
  readonly badge: string;
  /** The ⓘ tooltip: what a tool in this group is allowed to do. */
  readonly hint: string;
}

/** Wording for the four groups elitea-main classifies into. */
export function toolGroupDescriptor(groupId: string): ToolGroupDescriptor {
  switch (groupId) {
    case 'read':
      return {
        label: t('features.toolkits.toolGroups.read.label', 'Read'),
        badge: t('features.toolkits.toolGroups.read.badge', 'Read-only'),
        hint: t('features.toolkits.toolGroups.read.hint', 'Returns data. Nothing is created, changed or destroyed.'),
      };
    case 'create_update':
      return {
        label: t('features.toolkits.toolGroups.createUpdate.label', 'Create & update'),
        badge: t('features.toolkits.toolGroups.createUpdate.badge', 'Changes data'),
        hint: t('features.toolkits.toolGroups.createUpdate.hint', 'Creates or modifies data in the target system.'),
      };
    case 'delete':
      return {
        label: t('features.toolkits.toolGroups.delete.label', 'Delete'),
        badge: t('features.toolkits.toolGroups.delete.badge', 'Destructive'),
        hint: t('features.toolkits.toolGroups.delete.hint', 'Destroys data. Not reversible from Elitea.'),
      };
    case 'execute':
      return {
        label: t('features.toolkits.toolGroups.execute.label', 'Execute'),
        badge: t('features.toolkits.toolGroups.execute.badge', 'Unrestricted'),
        hint: t(
          'features.toolkits.toolGroups.execute.hint',
          'Runs a caller-supplied query, script, pipeline or raw API call. Effect is not bounded by the tool.',
        ),
      };
    default:
      /*
       * A group id this build has no wording for. It is rendered rather than
       * dropped: a served group nobody displays would silently hide every
       * tool in it from the picker, which is the "absence reads as
       * correctness" failure in its most expensive form — the tool would be
       * unselectable and nothing would say why.
       */
      return { label: groupId, badge: '', hint: '' };
  }
}

export interface ToolGroupSection extends ToolGroupDescriptor {
  readonly id: string;
  /** The tools this group holds AFTER the search filter, alphabetical by label. */
  readonly tools: readonly ToolActionOption[];
  /** EVERY tool name in the group, filter or no filter — what the header control toggles (ELITEA-2685 + ELITEA-2696). */
  readonly allToolValues: readonly string[];
  /** Selected out of TOTAL — both counted over the whole group, never over the filtered view (ELITEA-2690). */
  readonly selectedCount: number;
  readonly totalCount: number;
  /** True when every tool in the WHOLE group is selected — what the header control toggles off. */
  readonly allSelected: boolean;
}

export interface BuildToolGroupSectionsParams {
  readonly toolsOptions: readonly ToolActionOption[];
  /** Tool name → group id, as served. A tool missing from it has no group. */
  readonly toolGroups: Readonly<Record<string, string>> | undefined;
  /** The served fixed order. Falls back to `TOOL_GROUP_ORDER`. */
  readonly groupOrder: readonly string[] | undefined;
  readonly selectedTools: readonly string[];
  /** The search box's current text. Empty means no filtering. */
  readonly query: string;
}

export interface ToolGroupSections {
  /** Groups with at least one tool matching the query, in the fixed order. */
  readonly sections: readonly ToolGroupSection[];
  /** True when the toolkit has tools but the query matches none of them. */
  readonly noMatches: boolean;
}

/**
 * Matches a tool against the search text.
 *
 * BOTH the display name and the RAW tool name (ELITEA-2695): the chips read
 * "Read file" while the SDK, the agent prompt and every log line say
 * `read_file`, and a person who knows one spelling must not have to guess the
 * other. Case-insensitive, substring, no tokenisation — `read_f` has to find
 * `read_file`.
 */
function matchesQuery(option: ToolActionOption, query: string): boolean {
  if (query === '') return true;
  const needle = query.toLowerCase();
  return option.label.toLowerCase().includes(needle) || option.value.toLowerCase().includes(needle);
}

function compareByLabel(left: ToolActionOption, right: ToolActionOption): number {
  return left.label.localeCompare(right.label, undefined, { sensitivity: 'base' });
}

/**
 * Splits the available tools into the ordered, counted, filtered sections the
 * Tools section renders.
 *
 * A tool whose name is not in `toolGroups` falls into the FIRST group of the
 * order rather than being dropped — same reasoning as the unknown-group-id
 * case above: an unclassified tool must still be selectable. In practice this
 * does not arise for a served catalogue (elitea-main groups every tool it
 * serves, asserted in `tool_groups_test.go`); it arises for a toolkit whose
 * saved `available_mcp_tools` list has grown past the served classification.
 */
/** The bucketing half of `buildToolGroupSections`, its own function so that function stays inside the §3.5 complexity budget. */
function bucketByGroup(
  toolsOptions: readonly ToolActionOption[],
  toolGroups: Readonly<Record<string, string>> | undefined,
  fallbackGroup: string,
): { readonly byGroup: Map<string, ToolActionOption[]>; readonly encountered: readonly string[] } {
  const byGroup = new Map<string, ToolActionOption[]>();
  const encountered: string[] = [];
  for (const option of toolsOptions) {
    const groupId = toolGroups?.[option.value] ?? fallbackGroup;
    let bucket = byGroup.get(groupId);
    if (bucket === undefined) {
      bucket = [];
      byGroup.set(groupId, bucket);
      encountered.push(groupId);
    }
    bucket.push(option);
  }
  return { byGroup, encountered };
}

export function buildToolGroupSections(params: BuildToolGroupSectionsParams): ToolGroupSections {
  const { toolsOptions, toolGroups, groupOrder, selectedTools, query } = params;
  const order = groupOrder !== undefined && groupOrder.length > 0 ? groupOrder : TOOL_GROUP_ORDER;
  const fallbackGroup = order[0] ?? TOOL_GROUP_ORDER[0];

  const { byGroup, encountered } = bucketByGroup(toolsOptions, toolGroups, fallbackGroup);

  // The served order first, then anything served that this build's order does
  // not name — appended rather than dropped, for the same reason unknown ids
  // are rendered rather than dropped.
  const orderedIds = [...order, ...encountered.filter((id) => !order.includes(id))];

  const selected = new Set(selectedTools);
  const sections: ToolGroupSection[] = [];
  let anyMatch = false;
  for (const groupId of orderedIds) {
    const all = byGroup.get(groupId);
    if (all === undefined || all.length === 0) continue;
    const selectedCount = all.filter((option) => selected.has(option.value)).length;
    const tools = all.filter((option) => matchesQuery(option, query)).sort(compareByLabel);
    if (tools.length > 0) anyMatch = true;
    if (tools.length === 0) continue;
    sections.push({
      ...toolGroupDescriptor(groupId),
      id: groupId,
      tools,
      allToolValues: all.map((option) => option.value),
      selectedCount,
      totalCount: all.length,
      allSelected: selectedCount === all.length,
    });
  }

  return { sections, noMatches: toolsOptions.length > 0 && !anyMatch };
}

/**
 * The next `selected_tools` value after a group header is clicked.
 *
 * It toggles EVERY tool of the group, including the ones the search filter is
 * hiding, and touches nothing outside it (ELITEA-2685). Order is preserved for
 * the tools that stay: the stored array is the toolkit's saved value and
 * reshuffling it would show up as a spurious diff on every save.
 */
export function toggleGroupSelection(params: {
  readonly selectedTools: readonly string[];
  readonly groupTools: readonly string[];
  readonly select: boolean;
}): readonly string[] {
  const { selectedTools, groupTools, select } = params;
  if (!select) {
    const group = new Set(groupTools);
    return selectedTools.filter((tool) => !group.has(tool));
  }
  const already = new Set(selectedTools);
  const added = groupTools.filter((tool) => !already.has(tool));
  return added.length === 0 ? selectedTools : [...selectedTools, ...added];
}
